### DISCUSSION TITLE
Opinion of an External Independent Developer in July 2023: Add more Hooks :)

### DISCUSSION THREAD

**Discussion by webaxones** [woocommerce/woocommerce#44959] [WooCommerce, Blocks, and Site Editing]
Hello,
I am an independent WordPress website designer. I am both the graphic designer and the front and back-end developer.
I've been creating sites with WooCommerce for several years now.
I've been working with WooCommerce Blocks for many months.
The work you're doing is remarkable, and it's really a pity that you're not more numerous.
I'm taking the liberty of giving you an external opinion, even though I imagine the roadmap is already written :)
When we discuss amongst WooCommerce site developers, the phrase that always comes up is: why don't they put hooks everywhere?? (I'm talking about JavaScript).
I believe (and I'm not the only one) that you should prioritize the addition of hooks, more than features.
Why?
A single hook can allow an external developer to handle multiple features.
A hook is quicker for you to add than a feature. In an ideal world, a hook should be coded at the same time as a process, just like you don't come back later to internationalize your code.
While waiting for these features, no-code users could turn to developers. I want to clarify that I'm not defending this idea to make us indispensable, we have more than enough work. By adding more hooks, you would allow WooCommerce blocks to have more features quickly, even though they would initially be available through external developers, or plugins.
This mass addition of hooks would allow you to quantify the demand of each and thus establish priorities concerning the features to be offered on the blocks.
Finally, this would also reduce the number of issues since many of them would find their solution through hooks and less work on GitHub issues means more time for development.
I suspect that the whole problem is more complex than this, but at least you've had an external opinion :)
Thanks for reading me.
Loïc

**Comment by opr** [comment]
Hey Loïc thank you for taking the time to leave your feedback on WooCommerce Blocks, we really appreciate your input here!

> I'm taking the liberty of giving you an external opinion, even though I imagine the roadmap is already written :)

It is to some extent, but it could be possible that things are shifted around or goals changed if we see a really compelling reason for it, The opinions from merchants, shoppers, or developers using our product are super valuable in shaping the future roadmap.

> When we discuss amongst WooCommerce site developers, the phrase that always comes up is: why don't they put hooks everywhere?? (I'm talking about JavaScript).

This is something we see a lot, and the long and short of it is that anything we add to WooCommerce Blocks becomes another thing we have to maintain and another place where bugs could potentially appear. That's not to say hook are not useful, they are, but in some scenarios, a more explicit solution is better for both us as maintainers, and the developer consuming it. While having an explicit interface is also another chance for bugs to appear, we would prefer an interface to solve a specific problem, so we know exactly what the developers will be using it for reducing the time it takes to solve bugs. We now have the chance to introduce these new more explicit interfaces for extending WooCommerce, so we are trying our best to do this when we can.

> A single hook can allow an external developer to handle multiple features.

This is part of the issue, really. A hook can be used in many different ways, meaning we need to solve for various use-cases when adding them, and again more surface area for bugs to appear if two different developers decide to use the hooks for two completely different use-cases.

> This mass addition of hooks would allow you to quantify the demand of each and thus establish priorities concerning the features to be offered on the blocks.

I think it's safe to say the mass addition of hooks is **not** on the roadmap, and it is unlikely we will do that, it has been considered many times and we're pretty sure it's not the right way forward, as mentioned, instead we'll be focusing on the explicit interfaces to solve well-defined problems.

While this is just a feedback post, I'd like to ask if you have any specific examples of use cases that you feel a hook would solve? If you do, please feel free to let us know, and providing examples of how you solved it with hooks in the past would be super helpful! It might be the case that we have an extensibility point in place already that you could use, and if not it would help us identify what we need to work on.

Thanks again for taking the time to write to us.

**Reply by webaxones** [reply]
Thank you for responding.

I understand the logic of this decision, but I think it's a mistake.
Firstly, I want to address what you've said: if a third-party plugin generates bugs, it's not up to you to provide support.

What made WordPress successful are the hooks. By allowing it to be extended and transformed what was originally a blog engine into almost anything. Anything that works.
What made WooCommerce successful are also the hooks. Thanks to hooks, I've coded custom deposits, guarantees, rental management systems with such complex operations that only hooks could have allowed me to adapt the functionality of WooCommerce and WooCommerce Bookings.
You can't replace hooks with interfaces for a simple reason: because you can't judge the relevance of needs, not because you don't have the skills, but because this relevance varies depending on the origin of the request.

Example:
There's a PHP hook in WooCommerce that allows you to change the label of the Proceed to Checkout button. There is also a JS hook in the WooCommerce Cart block that allows you to change this label. This hook is mainly used for UX, marketing, or accessibility reasons.
Except that for France, it's different. It's used for legal reasons. You probably don't know this because a large majority of e-commerce site owners don't know it themselves, but the reality is the following: without this JS hook (or PHP one) the WooCommerce Cart block (or WooCommerce PHP template) is illegal in France. [The Hamon law of July 1, 2016](https://www.legifrance.gouv.fr/codes/article_lc/LEGIARTI000032226854) says:

"The professional ensures that the consumer, at the time of his order, explicitly acknowledges his payment obligation. To this end, the function used by the consumer to validate his order includes the clear and readable statement: order with payment obligation or a similar phrase, devoid of any ambiguity, indicating that placing an order entails its payment."

You might think: why not adapt the French translation of WooCommerce? Because the fr_FR locale covers several French-speaking countries, not just France, and none of them can impose a translation on the pretext of wanting to comply with its law. So, I therefore modify the translation of this button via this hook when the product is payable to: "Valider la commande et payer" (Validate the order and pay), which allows me to make my clients' shopping cart legal in France.

I could have cited all the examples of hooks I had to use on my WooCommerce sites, but I preferred to stick to this one which I find explicit, because it shows that you don't have the means to judge the relevance of a hook because the relationship between the hook and the need is not 1-1 but 1-n.

Personally, I think that if you persist in this view, the use of blocks will no longer be able to respond to all the cases we encounter, which will force us at best to code our own blocks, at worst to change tools. I'm not saying this in the form of a threat since I think we are in the same team, I'm just coldly looking at my needs, the possibilities offered to me, I'm comparing with my colleagues to see if I'm wrong or not, and I'm adapting my work and my tools accordingly.

:)

**Comment by nicomollet** [comment]
> This is part of the issue, really. A hook can be used in many different ways, meaning we need to solve for various use-cases when adding them, and again more surface area for bugs to appear if two different developers decide to use the hooks for two completely different use-cases.

This is what made the success of WordPress and WooCommerce.
Filters allow to create an infinity of plugins: "there is a plugin for that!".

By limiting customization, I fear it will drive way a good part of developers that won't see more benefits of the open source.
The huge WordPress ecosystem might disappear little by little by ignoring the usefulness of filters in the Gutenberg era.

If the "way forward" is create/duplicating common blocks for additional features / customization, it won't be that easy for developers and a good part of them will start looking elsewhere.

**Comment by marcusig** [comment]
Hi Thomas,

Thanks for giving us reasons on the current plan.
I'll express myself mostly about the cart and checkout blocks, as I haven't really used the other blocks too much so far. 
  
> This is part of the issue, really. A hook can be used in many different ways, meaning we need to solve for various use-cases when adding them, and again more surface area for bugs to appear if two different developers decide to use the hooks for two completely different use-cases.

While this makes sense for the WooCommerce devs and maintainers, it seems like an oversight on the fact that there is a large ecosystem that relies on being able to change things using hooks. 

The cart and checkout blocks being meant to replace their shortcode counterparts, for which dozens of plugins change the content using the PHP hooks. 

I develop and maintain plugins which interact with both checkout and cart, and at the moment I am not able to make those compatible, because I cannot modify the cart items without dodgy CSS.

For example those hooks: 

```
woocommerce_checkout_cart_item_visible
woocommerce_cart_item_visible
woocommerce_widget_cart_item_visible
woocommerce_order_item_visible
```

I can't currently prevent an item in the cart from being rendered. 
I could potentially hide it by adding a class and using CSS, but then it's still in the html, and an other plugin or theme could override that CSS. 

There are other issues which I am personally facing, such as: 

- the possibility to filter the data sent via the API (not just extending, also filtering the core data)
- the possibility to choose whether a sales badge is displayed or not
- the possibility to filter the product image (`woocommerce_cart_item_thumbnail`)
- the possibility to filter the formatting of cart item meta (the html is stripped in the blocks)

I'm happy to open issues to request specific hooks. If there are better solutions, links to specific documentation and examples on how to achieve those goals using best practices.

My concern with the case by case approach is the risk that compatibility could take a *long* time.

**Reply by nerrad** [reply]
> I'm happy to open issues to request specific hooks. If there are better solutions, links to specific documentation and examples on how to achieve those goals using best practices.

Please do open issues about the solutions you build for in the existing checkout flow using hooks. We may be able to preserve compatibility by bringing in a comparable hook, or it may be best solved by a different extensibility pattern. Either way, knowing the explicit problems you are solving for helps us ensure user needs are met.

**Reply by nerrad** [reply]
> My concern with the case-by-case approach is the risk that compatibility could take a long time.

Valid concern, however whenever a new extensibility gap comes up, we also consider how that same gap may solve multiple problems. Whatever extensibility pattern is released could then open the door to multiple uses/solutions. 

We prioritize what we know about :)

**Reply by marcusig** [reply]
Thanks for your feedback. I shall go through the various compatibility issues I face and open issues here.

**Comment by wpchannel** [comment]
Totally agree with my French colleagues, without lot of hooks we won't use WooCommerce anymore. Simple as that.

**Reply by ResponsiveFlow** [reply]
Do other platforms allow the ecommerce store to be complaint with French regulations?

**Reply by webaxones** [reply]
@tomxygen 
WooCommerce so far allowed it :)
But it's not just about regulations. The idea is that a hook can satisfy many needs. That these needs are so numerous, and essential for an ecommerce site, they cannot be managed by the plugin itself.

**Reply by ResponsiveFlow** [reply]
> WooCommerce so far allowed it 

Does it mean that they plan to stop supporting it?

**Reply by webaxones** [reply]
It's more complicated:
They want to reduce the number of hooks, to replace those that won't be reused with features under the editor.
But with one single hook, 10 different developers can meet 10 different needs. Therefore, there is a disagreement between the external developers, of which I am a part, and the developers of WooCommerce Blocks but also of Gutenberg :)

**Reply by nerrad** [reply]
I don't agree with the perception that our goal is to reduce the number of hooks. However, the default solution to an extensibility need isn't always going to be a hook.

**Comment by sebastienserre** [comment]
Without hooks... WooCommerce wouldn't exist... like all the plugins and WordPress never be what it is today.
Gui is good but for developers, hooks are better.

**Comment by webaxones** [comment]
Hello :)

I know that you are all very busy, but, @nerrad, could you clarify this concern we have regarding the hooks? 
You have a range of people here who work with WooCommerce throughout the year, be it as a web agency like @wpchannel or @nicomollet, a WooCommerce plugin developer like @marcusig (who hasn't received a response), a member of renowned plugins like Polylang (@sebastienserre), or even as a freelance developer like me? The non-replacement of all the hooks seems to us to be a dead end for creating ecommerce sites as we are currently doing. And these are not small sites. 
What do you have to say in response to the arguments we have put forward?
Thank you.
Loïc

**Comment by nerrad** [comment]
First, I just want to clarify that this thread and other threads like it are not ignored. As a group, we definitely are aware of the importance flexible interfaces and extensibility have been to the success of WordPress and WooCommerce. My own developer journey began with WordPress 1.5 so I personally know the value that the extensibility patterns in WordPress have meant for my own work with clients and custom products.

An important thing to keep in mind with the work we're explicitly doing in WooCommerce Blocks (and other newer interfaces across Woo) is that in some cases we won't be able to preserve the same extensibility pattern with previous functionality/feature sets because the underlying architecture and approach in WordPress has changed (or is changing). In some cases a PHP (or JS hook) is a great fit for extensibility, in others it isn't. There is also the tension we need to manage between the flexibility that is shifting from primarily code-based solutions to non-code-based, user-facing solutions where extensibility patterns also change. As much as possible we are also trying to preserve back-compat in the new things we build - or at least _keep_ existing functionality as an option for significantly custom work that may not have extensibility options yet in the new features/products/work.

Rather than prescribing a single approach/solution to all extensibility needs (eg hooks), we are very interested in finding the gaps in programmatic flexibility (what problems does it need to solve, both known and not known). It is important to not only offer extensibility solutions for those gaps but also hopefully improve the architecture so extensions and customizations built by developers have the flexibility needed but in a way that is easier to implement.

We're also looking at the opportunity to put into developer's hands interfaces that preserve and enhance a great user experience. Things like components, UI elements, and encapsulated error handling create pathways to building bespoke solutions for your clients and users in a way that is consistent with the rest of the interface and also adds more protection against plugin conflicts, errant edgecases etc.

How can you help us? Keep surfacing specific ways in which you have problems extending blocks or new interfaces. Try to be explicit about the problems you were able to solve with hooks. For example, instead of just writing something like "I used hook `woocommerce_x` before and need a replacement", include something like "I used hook `woocommerce_x` before to build solution y for my client." On our part, we're also making time to integrate various extensions to help discover some of these gaps ourselves and identify the most common extensibility needs we can prioritize addressing.

**Reply by webaxones** [reply]
Even if I know that this answer is not necessarily the one that everyone expects, I thank you for the frankness and the speed of your answer.
I just hope that the approach chosen by the team in charge of WordPress will not penalize this CMS that I appreciate.
:)

**Comment by eirichmond** [comment]
Indeed, an intriguing and important discussion regarding the compatibility of hooks and filters with the transition to blocks in WooCommerce. As a developer who uses hooks and filters, I acknowledge the significance of preserving backward compatibility while embracing the move to blocks.

WordPress has historically prioritised backward compatibility, ensuring that existing themes, plugins, and customizations continue to work even with updates. However, as the transition to blocks evolves, unforeseen issues may arise, surely raising concerns about maintaining this compatibility no?

Respecting hooks and filters from a classic development perspective is essential, as it enables smoother adoption of the block-based approach for both developers and users. By supporting these traditional mechanisms, it allows developers to retain their familiarity with the system and ensures a smoother migration path.

The potential impact on existing plugins and the overall user experience is a valid concern. Switching to block-based themes might cause some plugins in the ecosystem to stop working if they are not adapted to the new paradigm. Store owners relying on these plugins could face challenges if their essential functionalities are disrupted, what would that say about user experience?

I for one will be following this discussion and hope to surface my own views as and when I have a case example, thanks to @webaxones and other contributions for bring this to the attention of others.
